File: java/PasswordChecker.java
import java.util.Scanner;

public class PasswordChecker {

    // main(String[] args)
    // Behavior: Entry point for the program; prompts user for password input, calculates strength score, and displays both score and strength rating
    // Parameters: args - command line arguments (not used)
    // Returns: void
    // Exceptions: None
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter a password: ");
        String password = scanner.nextLine();

        int score = calculateScore(password);
        String strength = getStrength(score);

        System.out.println("Score: " + score);
        System.out.println("Strength: " + strength);

        scanner.close();
    }

    /**
     * calculateScore(String password)
     * Behavior: Evaluates password strength by checking five criteria: minimum length (8 chars), presence of lowercase, uppercase, digit, and special character. Increments score for each criterion met.
     * Parameters: password - the password string to evaluate
     * Returns: int score between 0-5, where each point represents one criterion met
     * Exceptions: None
     */
    public static int calculateScore(String password) {
        int score = 0;

        if (password.length() >= 8) {
            score++;
        }
        if (hasLowercase(password)) {
            score++;
        }
        if (hasUppercase(password)) {
            score++;
        }
        if (hasDigit(password)) {
            score++;
        }
        if (hasSpecialChar(password)) {
            score++;
        }

        return score;
    }

    /**
     * getStrength(int score)
     * Behavior: Maps a numeric score to a human-readable strength level. Score 0-2 returns "Weak", 3-4 returns "Moderate", 5 returns "Strong".
     * Parameters: score - integer value representing password strength (typically 0-5)
     * Returns: String containing strength level ("Weak", "Moderate", or "Strong")
     * Exceptions: None
     */
    public static String getStrength(int score) {
        if (score <= 2) {
            return "Weak";
        } else if (score <= 4) {
            return "Moderate";
        } else {
            return "Strong";
        }
    }

    // hasLowercase(String password)
    // Behavior: Iterates through each character in the password to check for the presence of lowercase letters
    // Parameters: password - the password string to check
    // Returns: true if at least one lowercase letter is found, false otherwise
    // Exceptions: None
    private static boolean hasLowercase(String password) {
        for (char c : password.toCharArray()) {
            if (Character.isLowerCase(c)) {
                return true;
            }
        }
        return false;
    }

    // hasUppercase(String password)
    // Behavior: Iterates through each character in the password to check for the presence of uppercase letters
    // Parameters: password - the password string to check
    // Returns: true if at least one uppercase letter is found, false otherwise
    // Exceptions: None
    private static boolean hasUppercase(String password) {
        for (char c : password.toCharArray()) {
            if (Character.isUpperCase(c)) {
                return true;
            }
        }
        return false;
    }

    // hasDigit(String password)
    // Behavior: Iterates through each character in the password to check for the presence of numeric digits
    // Parameters: password - the password string to check
    // Returns: true if at least one digit is found, false otherwise
    // Exceptions: None
    private static boolean hasDigit(String password) {
        for (

File: c/password_checker.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

// has_lowercase(char *password)
// Behavior: Scans the password string to check for the presence of lowercase letters
// Parameters: password - pointer to null-terminated string to check
// Returns: 1 if at least one lowercase letter is found, 0 otherwise
// Exceptions: None
int has_lowercase(char *password) {
    for (int i = 0; password[i]; i++) {
        if (islower(password[i])) {
            return 1;
        }
    }
    return 0;
}

// has_uppercase(char *password)
// Behavior: Scans the password string to check for the presence of uppercase letters
// Parameters: password - pointer to null-terminated string to check
// Returns: 1 if at least one uppercase letter is found, 0 otherwise
// Exceptions: None
int has_uppercase(char *password) {
    for (int i = 0; password[i]; i++) {
        if (isupper(password[i])) {
            return 1;
        }
    }
    return 0;
}

// has_digit(char *password)
// Behavior: Scans the password string to check for the presence of numeric digits
// Parameters: password - pointer to null-terminated string to check
// Returns: 1 if at least one digit is found, 0 otherwise
// Exceptions: None
int has_digit(char *password) {
    for (int i = 0; password[i]; i++) {
        if (isdigit(password[i])) {
            return 1;
        }
    }
    return 0;
}

// has_special(char *password)
// Behavior: Scans the password string to check for the presence of special characters (non-alphanumeric)
// Parameters: password - pointer to null-terminated string to check
// Returns: 1 if at least one special character is found, 0 otherwise
// Exceptions: None
int has_special(char *password) {
    for (int i = 0; password[i]; i++) {
        if (!isalnum(password[i])) {
            return 1;
        }
    }
    return 0;
}

// main()
// Behavior: Prompts user for password input, evaluates password strength based on five criteria (length, lowercase, uppercase, digit, special char), and displays strength score and rating
// Parameters: None
// Returns: 0 on successful execution
// Exceptions: None
int main() {
    char password[256];
    int score = 0;

    printf("Enter a password: ");
    fgets(password, sizeof(password), stdin);

    // Remove newline from fgets
    password[strcspn(password, "\n")] = '\0';

    if (strlen(password) >= 8) score++;
    if (has_lowercase(password)) score++;
    if (has_uppercase(password)) score++;
    if (has_digit(password)) score++;
    if (has_special(password)) score++;

    printf("Score: %d\n", score);

    if (score <= 2) {
        printf("Strength: Weak\n");
    } else if (score <= 4) {
        printf("Strength: Moderate\n");
    } else {
        printf("Strength: Strong\n");
    }

    return 0;
}
